Implement `/debug/archive` handler

sachinpuranik opened the pull request:
Closes #3584.

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [ ] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -37,11 +41,23 @@ import (
 )
 
+var (
+	graphicsPath = "/debug/graphs"
```
```suggestion
	graphsPath = "/debug/graphs"
```

sachinpuranik replied:
done

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -37,11 +41,23 @@ import (
+	varsPath     = "/debug/vars"
+)
+
+func generateFileName(prefix, filename string) string {
```
I don't think it is worth extracting that function

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -123,3 +142,80 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
 	l.Sugar().Info("Debug server stopped.")
 }
+
+func getArchiveHandler() http.HandlerFunc {
```
https://github.com/FerretDB/FerretDB/blob/b9eafc581920ab30b1546cc46141486ee9945a1a/CONTRIBUTING.md?plain=1#L263

sachinpuranik replied:
done. added comments.

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -123,3 +142,80 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
 }
+
+func getArchiveHandler() http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
```
I think the function itself could have that signature instead of returning a function with that signature

sachinpuranik replied:
done.

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -123,3 +142,80 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
+
+func getArchiveHandler() http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		var (
+			scheme          string
+			u               url.URL
+			responses       = make(map[string][]byte, 2)
+			debugFilePrefix = "FerretDB"
+			debugFileName   = "debug.zip"
+			err             error
+		)
```
I think we should define variable when we need them, not in advance

sachinpuranik replied:
ok.

sachinpuranik replied:
addressed.

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -123,3 +142,80 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
+		u.Path = metricsPath
+		u.Host = req.Host
+
+		if req.URL.Scheme == "" {
```
It can't be anything other than "http"

sachinpuranik replied:
During testing I found that scheme was coming as "", hence this was required.
Any other suggestion?

sachinpuranik replied:
@AlekSi , WDYT?

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -123,3 +142,80 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
```
There should not be a need to read all data into a slice and then write this slice. We have a reader there; we use `io.Copy` above that could accept that reader.

sachinpuranik replied:
ok. Optimised in new impl.

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -123,3 +142,80 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
+			}
+			_, err = io.Copy(fileWriter, bytes.NewReader(response))
+			if err != nil {
+				http.Error(rw, fmt.Sprintf("failed - adding %s to zip , (%s)", fileName, err.Error()), http.StatusInternalServerError)
```
I think we should log error like that, but don't return them

sachinpuranik replied:
we can do that. But since this is specifically for debugging , isn't it good to return an error , so for any issues , error is clearly known. it shall not be lost in the logs. WDYT?

sachinpuranik replied:
@AlekSi , WDYT?

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -123,3 +142,80 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
+			return
+		}
+
+		u.Path = pprofPath
+		pprofFileName := generateFileName(debugFilePrefix, filepath.Base(pprofPath)+".html")
+		responses[pprofFileName], err = performRequest(u)
```
We should get a memory profile, not HTML page with a list of profiles

AlekSi replied:
I'm not sure why we decided to parse HTML, but we definitely should not do that. The URL for the memory profile is fixed: `debug/pprof/heap` (without query parameters).

Also, please don't resolve conversations: https://github.com/FerretDB/FerretDB/blob/c8f8c6851e4a07ae5cc9297ae2a5019bcea6c7f1/CONTRIBUTING.md?plain=1#L358-L362

It might help to re-read the whole file.

sachinpuranik replied:
Hi @AlekSi  ,HTML parsing is required since we add dynamic resource tracking in pprof, I am assuming we need them as well in zip. if not , we should improve the task by specifying what's exactly required.  Additionally Debug=1 , exports the data in the legacy text format. I tried exporting few files without this option and they were binary , some non readable info.

sachinpuranik replied:
Do we only one URL - "debug/pprof/heap" to be grabbed and nothing else.

AlekSi replied:
> we add dynamic resource tracking in pprof

Why do we need that?

> Do we only one URL - "debug/pprof/heap" to be grabbed and nothing else.

Yes.

> Additionally Debug=1 , exports the data in the legacy text format. I tried exporting few files without this option and they were binary , some non readable info.

That format is readable with [`go tool pprof`](https://go.dev/doc/diagnostics#profiling) and contains more information than the legacy text format. We need that.

sachinpuranik replied:
Sure. Updated the code to gather only two files in zip. matrics and heap. test updated accordingly.

AlekSi replied:
@noisersup PTAL ^ We don't need `debug=1`

AlekSi commented on internal/util/debug/debug_test.go:
```diff
@@ -14,8 +14,83 @@
+
+	l := zap.S()
+
+	go RunHandler(ctx, host, metricsRegisterer, l.Named("debug").Desugar())
```
That goroutine is never stopped

sachinpuranik replied:
It's a test and thread shall stop automatically as soon as test ends. Do we need explicit stop?

sachinpuranik replied:
WDYT?

henvic replied:
I'd gracefully stop by setting a cancellable context and calling `cancel()` from a `t.Cleanup()` just to avoid any log messages that might confuse me.

sachinpuranik commented:
Hi Team , PTAL.
@AlekSi @henvic

sachinpuranik commented:
Question - on the pprof page we have two links... Do we want both the files to come in the archive?
http://127.0.0.1:8088/debug/pprof/goroutine?debug=1
http://127.0.0.1:8088/debug/pprof/goroutine?debug=2

AlekSi reviewed (changes requested):
https://github.com/FerretDB/FerretDB/pull/3895#discussion_r1441993707

sachinpuranik commented:
request your comments on other places too.

AlekSi reviewed (commented):
No amount of merging with the `main` branch will make CI pass :)

sachinpuranik commented:
Hi @AlekSi, I was more of focusing on approach and core implementation.
Updated the go.mod and fixed the issue.
Though "Go / sqlite 3/3 / Run (pull_request) " Still fails due to some cursor test fail. It's unrelated to change ?
How do we tackle it.

AlekSi commented on internal/util/debug/archive_test.go:
```diff
@@ -0,0 +1,106 @@
+	"github.com/FerretDB/FerretDB/internal/util/state"
+)
+
+func findInStringsArray(findStr string, list []string) bool {
```
https://pkg.go.dev/slices@go1.21.6#Contains

sachinpuranik replied:
done.

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,125 @@
+	rw.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%s", zipName))
+
+	zipWriter := zip.NewWriter(rw)
+	defer zipWriter.Close() //nolint:errcheck // we are only reading it
```
I don't think we are only reading _the writer_. It is not safe to ignore that error.

sachinpuranik replied:
Done.
modified the code , if error comes now we will return the internal error.
Also lint bypass was not needed post modification. so removed.

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,125 @@
+	"path/filepath"
+)
+
+type archiveUrlList []url.URL
```
https://github.com/FerretDB/FerretDB/blob/8d37e8cdc915593107c4eba3f02c50bba70a0e45/CONTRIBUTING.md?plain=1#L263

That's everywhere. Please re-read out contributing documentation.

sachinpuranik replied:
done.

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,125 @@
+
+type archiveUrlList []url.URL
+
+// archiveHandler - is a handler for creating the archive containing all the debug and metrics info.
```
```suggestion
// archiveHandler is a handler for creating the archive containing all the debug and metrics info.
```

To follow all other declaration comments.

But that comment is also incorrect – that archive does not contain all debug and metrics info.

sachinpuranik replied:
Updated the sentence. This was added when i was adding all the files to zip.

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,125 @@
+
+// archiveHandler - is a handler for creating the archive containing all the debug and metrics info.
+func archiveHandler(rw http.ResponseWriter, req *http.Request) {
+	var (
+		scheme  string
+		urlList archiveUrlList
+	)
+
+	urlList = make(archiveUrlList, 0)
```
https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_use_a_consistent_declaration_style

(the page might be useful)

sachinpuranik replied:
ok. Made it uniform.

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,125 @@
+
+	urlList = make(archiveUrlList, 0)
+
+	u := new(url.URL)
+	u.Path = metricsPath
+	u.Host = req.Host
+
+	if req.URL.Scheme == "" {
+		scheme = "http"
+	}
+
+	u.Scheme = scheme
+	urlList = append(urlList, *u)
```
I'm sorry, but that's a very complicated way. Just use `u := &url.URL{...}`

sachinpuranik replied:
simplified the usage.

sachinpuranik replied:
Finding : 
Linter shows a message as below for usage of &url . That was the reason I made it like that earlier.

./internal/util/debug/archive.go:35:7: zero value ptr alloc: use new(T) for *T allocation
./internal/util/debug/archive.go:46:6: zero value ptr alloc: use new(T) for *T allocation

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,125 @@
+
+	rw.Header().Set("Content-Type", "application/zip")
+
+	var (
+		debugFilePrefix = "FerretDB"
+		debugFileName   = "debug.zip"
+	)
+
+	zipName := fmt.Sprintf("%s-%s", debugFilePrefix, debugFileName)
```
```suggestion
	zipName := "FerretDB-debug.zip"
```

In general, that's the biggest problem in that PR – it is way too complicated for simple things like that one. Simplify. Then simplify some more :)

sachinpuranik replied:
simplified.

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -37,11 +37,20 @@ import (
 )
 
+var (
+	graphsPath  = "/debug/graphs"
```
Why do we have those variables and why they are variables?

sachinpuranik replied:
Changed to constants. And used these constants everywhere in the code. Some if the URL are used in test as well. hence converted all the URL to constants and used uniformly across.

sachinpuranik commented:
HI @AlekSi , do we have a way to trigger failed CI - steps again?

chilagrow commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// addFileToArchive function responsible for adding content to zip file.
+func addFileToArchive(fileName string, fileReader io.ReadCloser, zipWriter *zip.Writer) error {
```
Do we need this separate function @noisersup? :thinking:

AlekSi replied:
I guess yeah, for `defer` to work

chilagrow commented on internal/util/debug/archive_test.go:
```diff
@@ -0,0 +1,96 @@
+
+	for _, file := range zipReader.File {
+		t.Logf("\nverifying file : %s", file.Name)
+		require.Equal(t, true, slices.Contains(fileList, file.FileHeader.Name), "file should be present in archive")
```
Shall we check file is not empty?

chilagrow reviewed (commented):
Just minor comments @noisersup

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,91 @@
+		metricsPath,
+		pprofPath + "/heap",
+	} {
+		values := url.Values{}
+		values.Add("debug", "1")
+
+		u := url.URL{
+			Scheme:   "http",
+			Host:     req.Host,
+			Path:     path,
+			RawQuery: values.Encode(),
+		}
```
I still don't understand why we make it so hard.

```
uri := fmt.Sprintf("http://%s%s", addr, path)
```

`addr` is should come from the listener (`req.Host` is just not correct when, for example, a reverse proxy is present).

And we don't need `debug=1` as mentioned elsewhere.

The same in the test.

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,91 @@
+			RawQuery: values.Encode(),
+		}
+
+		req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+		if err != nil {
+			http.Error(rw, lazyerrors.Error(err).Error(), http.StatusInternalServerError)
+			return
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			http.Error(rw, lazyerrors.Error(err).Error(), http.StatusInternalServerError)
+			return
+		}
```
All that is just `http.Get`.

The same in the test

AlekSi commented on internal/util/debug/archive.go:
```diff
@@ -0,0 +1,91 @@
+func addFileToArchive(fileName string, fileReader io.ReadCloser, zipWriter *zip.Writer) error {
+	defer fileReader.Close() //nolint:errcheck // we are only reading it
+
+	fileWriter, err := zipWriter.Create(fileName)
```
```suggestion
	f, err := zipWriter.Create(fileName)
```

https://go.dev/wiki/CodeReviewComments#variable-names

AlekSi commented on internal/util/debug/archive_test.go:
```diff
@@ -0,0 +1,96 @@
+		"metrics", "heap",
+	}
+
+	host := "127.0.0.1:5454"
```
That will fail if the port is occupied by something else.

Listening on `0.0.0.0:0` never fails.

AlekSi commented on internal/util/debug/archive_test.go:
```diff
@@ -0,0 +1,96 @@
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	go RunHandler(ctx, host, metricsRegisterer, l.Named("debug").Desugar())
+
+	// Wait for the server to start
+	time.Sleep(time.Second)
```
Not good enough

AlekSi commented on internal/util/debug/archive_test.go:
```diff
@@ -0,0 +1,96 @@
+	l := zap.S()
+
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	go RunHandler(ctx, host, metricsRegisterer, l.Named("debug").Desugar())
```
We don't want to leak that goroutine

AlekSi commented on internal/util/debug/archive_test.go:
```diff
@@ -0,0 +1,96 @@
+	require.Equal(t, len(fileList), len(zipReader.File), "needs to be same as length of fileList")
+
+	for _, file := range zipReader.File {
+		t.Logf("\nverifying file : %s", file.Name)
```
`\n`

chilagrow reviewed (dismissed):
One minor comment :hugs:

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -88,6 +91,29 @@ func Listen(opts *ListenOpts) (*Handler, error) {
 
 	stdL := must.NotFail(zap.NewStdLogAt(opts.L, zap.WarnLevel))
 
+	archiveDurations := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "archive_response_seconds",
+			Help:      "Archive response time seconds.",
+			Buckets:   []float64{0.1, 0.5, 1, 5},
+		},
+		[]string{"code"},
+	)
```
I don't see how that would be useful given that amount of information in the archive will change over time

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -97,25 +123,98 @@ func Listen(opts *ListenOpts) (*Handler, error) {
+			defer func() {
+				if err := zipWriter.Close(); err != nil {
+					opts.L.Error("Archive handler failed", zap.Error(err))
+					rw.WriteHeader(http.StatusInternalServerError)
```
It is not possible to send a header with status after some content was already sent

noisersup replied:
My bad, fixed

AlekSi reviewed (commented):
That code should use slog, not zap

chilagrow commented on internal/util/debug/debug.go:
```diff
@@ -97,25 +123,98 @@ func Listen(opts *ListenOpts) (*Handler, error) {
+			scrapeReq := must.NotFail(http.NewRequestWithContext(
+				ctx,
+				http.MethodGet,
+				fmt.Sprintf("http://%s%s", req.Host, "/debug/metrics"),
```
Why not use https://pkg.go.dev/net/url#JoinPath?

noisersup replied:
We don't handle user provided paths here, path traversal is not a problem here.
We can use `JoinPath` here nevertheless, but the code'll be just less simple:
```
			must.NotFail(url.JoinPath("http://"+req.Host, "/debug/metrics")),
```

Although I think what could be done is to not use `req.Host` at all. It doesn't make sense to use and potentially might be harmful, thank you for noticing the line!

chilagrow replied:
Ok. What is harmful though?

noisersup replied:
The `Host` field is not validated by the server, and it can be manipulated by the client. That might lead to things such as forcing the server to query user-specified server, or something unexpected and more harmful.

AlekSi replied:
🏅

chilagrow commented on internal/util/debug/debug.go:
```diff
@@ -97,25 +123,98 @@ func Listen(opts *ListenOpts) (*Handler, error) {
+
+			_, err = io.Copy(metricsFile, resp.Body)
+
+			_ = resp.Body.Close()
```
Error from Close() should be logged right?

noisersup replied:
Not really, it returns error to conform to `Closer` interface. The `net/http` calls `Response.Body.Close` without handling error in multiple places: [1](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/net/http/response.go;l=338), [2](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/net/http/client.go;l=708).

In `net/http` docs there's an example with `resp.Body.Close()` not being checked: https://pkg.go.dev/net/http#pkg-overview.

Also PTAL at discussion here: securego/gosec#925.

Overall I didn't check for error, because in previous approach I've used deferred close. I'd suggest to keep it like that to not introduce another 3-5 lines after each Close to handle error that doesn't occur. It'll be also less confusing for the future if someone might want to use deferred call here.

chilagrow commented on internal/util/debug/debug.go:
```diff
@@ -97,25 +123,98 @@ func Listen(opts *ListenOpts) (*Handler, error) {
+			}
+
+			_, err = io.Copy(heapFile, resp.Body)
+			_ = resp.Body.Close()
```
Same here, error from Close should be logged

noisersup replied:
https://github.com/FerretDB/FerretDB/pull/3895#discussion_r1684277509

chilagrow commented:
> That code should use slog, not zap

@noisersup this comment was missed.

chilagrow commented on internal/util/debug/debug_test.go:
```diff
@@ -58,21 +62,73 @@ func TestProbes(t *testing.T) {
+			require.NoError(t, err)
+
+			assert.Equal(t, 1, n, "file should contain any data, but was empty")
+			require.NoError(t, f.Close())
```
Can you put `f.Close` closer to  `file.Open()` using `t.Cleanup`. Not all the code path reaches here.

noisersup replied:
done

rumyantseva commented on internal/util/debug/debug.go:
```diff
@@ -100,25 +116,84 @@ func Listen(opts *ListenOpts) (*Handler, error) {
+		defer func() {
+			if err := zipWriter.Close(); err != nil {
+				l.ErrorContext(ctx, "Archive handler failed", logging.Error(err))
+				return
```
I'm not sure if this `return` is needed, as we are already inside `defer` :)

rumyantseva commented on internal/util/debug/debug.go:
```diff
@@ -100,25 +116,84 @@ func Listen(opts *ListenOpts) (*Handler, error) {
+		metricsFile, err := zipWriter.Create("metrics")
+		if err != nil {
+			l.ErrorContext(ctx, "Archive handler failed", logging.Error(err))
+			return
```
What does the response to the user look like in this case and all the other cases when an error is returned? An empty zip file?

As this is a debug handler, I don't think it's super important to return something particular, but on the other hand, it's nice if the response is clear (like `500` or something). However, I'm not sure if it is possible to return something like 500 page, as we have already set some content-type headers. So, this is not critical, just some thinking.

noisersup replied:
The request will just end with StatusOK. I think we could use multipart here to handle error properly, but I'd suggest doing that in separate PR. @FerretDB/core WDYT?

noisersup replied:
and about multipart? :)

AlekSi replied:
I don't know how one can send a file using the multipart encoding, but I know for sure that nothing would handle that response. Definitely not `curl` or browser that would be used to download that debug archive.

rumyantseva reviewed (approved):
Looks good to me, asked a couple of questions for my understanding.

AlekSi commented on internal/util/debug/debug.go:
```diff
@@ -100,25 +116,84 @@ func Listen(opts *ListenOpts) (*Handler, error) {
+			return
+		}
+
+		debugAddr := ctx.Value(http.LocalAddrContextKey).(net.Addr)
```
Oooh, that's clever
